tests/unit/test_gcp_clients.py

```diff
@@ -0,0 +1,256 @@
+        """Client uses multiple filters to process results."""
+        caplog.set_level(logging.DEBUG)
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+
+        if instance_meta:
+            instance_dict = compute_rsp['items']['us-west1-z']['instances'][0]
+            blacklisted_instance = copy.deepcopy(instance_dict)
+            blacklisted_instance['name'] = 'instance-2'
+            compute_rsp['items']['us-west1-z']['instances'].append(
+                blacklisted_instance)
+
+        if isinstance(instance_meta, list):
+            blacklisted_instance['tags']['items'] = instance_meta
+            gce_client.blacklisted_tags.append(instance_meta[0])
+        elif isinstance(instance_meta, dict):
+            blacklisted_instance['metadata']['items'].append(instance_meta)
+            blacklisted_metadata = {
+                instance_meta['key']: instance_meta['value']
+            }
+            gce_client.blacklisted_metadata.append(blacklisted_metadata)
+
+        with aioresponses() as m:
+            filter_url = (f'{patch_compute_base_url}{query_str}')
+            m.get(filter_url, payload=compute_rsp)
+
+            kwargs = {}
+            if instance_meta is None:
+                kwargs = {'instance_filter': query_str[query_str.find('zone'):]}
+
+            results = await gce_client.list_instances(
+                'test-project',
+                page_size=10,
+                **kwargs)
+
+        expected_results = [{
+            'hostname': 'instance-1',
+            'internal_ip': '192.168.0.1',
+            'external_ip': '1.1.1.1'
+        }]
+        assert expected_results == results
+        assert log_call_count == len(caplog.records)
+
+    @pytest.mark.asyncio
+    async def test_list_instances_bad_json(
+            self, compute_rsp, patch_compute_base_url, get_gce_client, caplog):
+        """Client ignores incomplete API replies."""
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+        caplog.set_level(logging.DEBUG)
+        del compute_rsp['items']['us-west1-z']['instances'][0]['name']
+        with aioresponses() as m:
+            filter_url = f'{patch_compute_base_url}maxResults=10&filter='
+            m.get(filter_url, payload=compute_rsp)
+
+            results = await gce_client.list_instances(
+                'test-project', page_size=10)
+
+        assert [] == results
+        assert 3 == len(caplog.records)
+
+    @pytest.mark.asyncio
+    async def test_list_instances_retrieves_multiple_pages(
+            self, compute_rsp, patch_compute_base_url, get_gce_client, caplog):
+        """Client successfully retrieves multiple pages of results from API."""
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+        page2 = copy.deepcopy(compute_rsp)
+        page2['items']['us-west1-z']['instances'][0]['name'] = 'instance-2'
+        compute_rsp['nextPageToken'] = '123token123'
+        with aioresponses() as m:
```

Comment:
Maybe assert two requests here too like you did for the list projects one above?